folly fails to build - Histogram.cpp

Apologies for yet another folly issue.

TLDR:
```
E:\git\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\stats\Histogram.cpp(42): error C5037: 'folly::detail::HistogramBuckets<__int64,folly::detail::Bucket<T>>::getPercentileBucketIdx': an out-of-line definition of a member of a class template cannot have default arguments
        with
        [
            T=int64_t
        ]
```

```

Call Stack (most recent call first):
  E:/git/vcpkg/scripts/cmake/vcpkg_install_cmake.cmake:49 (vcpkg_execute_required_process)
  E:/git/vcpkg/ports/folly/portfile.cmake:41 (vcpkg_install_cmake)
  E:/git/vcpkg/scripts/ports.cmake:72 (include)


Error: Building package folly:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: folly:x64-windows
  Vcpkg version: 0.0.90-2017-10-04-d8507cd159f8de6ae174cae9894c926f75fe50d7
```

To actually obtain an error, it was necessary to run cmake manually, as no errors appeared in the log files.

```
E:\git\vcpkg\buildtrees\folly>cd E:/git/vcpkg/buildtrees/folly/x64-windows-rel

E:\git\vcpkg\buildtrees\folly\x64-windows-rel>E:\git\vcpkg\downloads\cmake-3.9.3-win32-x86\bin\cmake.exe --build . --config Release --target install -- -v

...

FAILED: CMakeFiles/folly_base.dir/folly/stats/Histogram.cpp.obj
C:\PROGRA~2\MIB055~1\2017\ENTERP~1\VC\Tools\MSVC\1411~1.255\bin\HostX64\x64\cl.exe   /TP -DWIN32_LEAN_AND_MEAN -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_WARNINGS -D_HAS_AUTO_PTR_ETC=1 -D_SCL_SECURE_NO_WARNINGS -D_STL_EXTRA_DISABLED_WARNINGS="4774 4987" -IE:\git\vcpkg\buildtrees\folly\src\folly-2017.07.17.01 -I. -IE:\git\vcpkg\installed\x64-windows\include /DWIN32 /D_WINDOWS /W3 /utf-8 /GR  /MP /MD /O2 /Oi /Gy /DNDEBUG /Z7   /EHa /GF /Zc:referenceBinding /Zc:rvalueCast /Zc:implicitNoexcept /Zc:strictStrings /Zc:threadSafeInit /Zc:throwingNew /std:c++latest /permissive- /bigobj /favor:blend /Zc:inline /Wall /MP /Gw /Gy /Qpar /Oi /Ot /wd4191 /wd4291 /wd4309 /wd4310 /wd4366 /wd4587 /wd4592 /wd4723 /wd4724 /wd4868 /wd4996 /wd4068 /wd4091 /wd4146 /wd4800 /wd4018 /wd4365 /wd4388 /wd4389 /wd4100 /wd4459 /wd4505 /wd4701 /wd4702 /wd4061 /wd4127 /wd4200 /wd4201 /wd4296 /wd4316 /wd4324 /wd4355 /wd4371 /wd4435 /wd4514 /wd4548 /wd4574 /wd4582 /wd4583 /wd4619 /wd4623 /wd4625 /wd4626 /wd4647 /wd4668 /wd4706 /wd4710 /wd4711 /wd4714 /wd4820 /wd5026 /wd5027 /wd5031 /we4099 /we4129 /we4566 /showIncludes /FoCMakeFiles\folly_base.dir\folly\stats\Histogram.cpp.obj /FdCMakeFiles\folly_base.dir\ /FS -c E:\git\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\stats\Histogram.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.11.25507.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

E:\git\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\stats\Histogram.cpp(42): error C5037: 'folly::detail::HistogramBuckets<__int64,folly::detail::Bucket<T>>::getPercentileBucketIdx': an out-of-line definition of a member of a class template cannot have default arguments
        with
        [
            T=int64_t
        ]
```

Same issue, although I did find stuff in logs. Prior to the histogram related error are these:

> C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\io\async\AsyncSocket.cpp(1589): error C2039: 'Control': is not a member of 'msghdr'
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly/portability/Sockets.h(42): note: see declaration of 'msghdr'
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\io\async\AsyncSocket.cpp(1589): error C2228: left of '.len' must have class/struct/union
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\io\async\AsyncSocket.cpp(1589): error C2228: left of '.buf' must have class/struct/union
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\io\async\AsyncSocket.cpp(1593): error C2039: 'Control': is not a member of 'msghdr'
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly/portability/Sockets.h(42): note: see declaration of 'msghdr'
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\io\async\AsyncSocket.cpp(1593): error C2228: left of '.len' must have class/struct/union
C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\io\async\AsyncSocket.cpp(1593): error C2228: left of '.buf' must have class/struct/union

Much further down in the log is

> C:\src\vcpkg\buildtrees\folly\src\folly-2017.07.17.01\folly\stats\Histogram.cpp(42): error C5037: 'folly::detail::HistogramBuckets<__int64,folly::detail::Bucket<T>>::getPercentileBucketIdx': an out-of-line definition of a member of a class template cannot have default arguments
        with
        [
            T=int64_t
        ]

--- a/ports/folly/CONTROL
+++ b/ports/folly/CONTROL
@@ -1,4 +1,4 @@
 Source: folly
-Version: v2017.07.17.01-1
+Version: v2017.10.02.00
 Description: An open-source C++ library developed and used at Facebook. The library is UNSTABLE on Windows
 Build-Depends: zlib, openssl, boost, libevent, double-conversion, glog, gflags, lz4, liblzma, snappy

--- a/ports/folly/fix-MSG_ERRQUEUE.diff
+++ b/ports/folly/fix-MSG_ERRQUEUE.diff
@@ -1,0 +1,22 @@
+diff --git a/folly/io/async/AsyncSocket.cpp b/folly/io/async/AsyncSocket.cpp
+index 7f8c5f13..f77adbc0 100644
+--- a/folly/io/async/AsyncSocket.cpp
++++ b/folly/io/async/AsyncSocket.cpp
+@@ -41,7 +41,7 @@ namespace fsp = folly::portability::sockets;
+ namespace folly {
+ 
+ static constexpr bool msgErrQueueSupported =
+-#ifdef MSG_ERRQUEUE
++#if defined(MSG_ERRQUEUE) && !defined(_WIN32)
+     true;
+ #else
+     false;
+@@ -1551,7 +1551,7 @@ void AsyncSocket::handleErrMessages() noexcept {
+     return;
+   }
+ 
+-#ifdef MSG_ERRQUEUE
++#if defined(MSG_ERRQUEUE) && !defined(_WIN32)
+   uint8_t ctrl[1024];
+   unsigned char data;
+   struct msghdr msg;

--- a/ports/folly/fix-histogram.diff
+++ b/ports/folly/fix-histogram.diff
@@ -1,0 +1,11 @@
+--- a/folly/stats/Histogram.h
++++ b/folly/stats/Histogram.h
+@@ -485,7 +485,7 @@
+ // with /permissive- (the default for the CMake build), so we directly include
+ // the -defs as if it were -inl, and don't provide the explicit instantiations.
+ // https://developercommunity.visualstudio.com/content/problem/81223/incorrect-error-c5037-with-permissive.html
+-#if defined(_MSC_VER) && _MSC_FULL_VER >= 191125506 && _MSC_FULL_VER < 191125542
++#if defined(_MSC_VER) && _MSC_FULL_VER >= 191125506 && _MSC_FULL_VER <= 191125547
+ #define FOLLY_MSVC_USE_WORKAROUND_FOR_C5037 1
+ #else
+ #define FOLLY_MSVC_USE_WORKAROUND_FOR_C5037 0

--- a/ports/folly/fix-malloc.diff
+++ b/ports/folly/fix-malloc.diff
@@ -1,0 +1,12 @@
+--- a/folly/Malloc.h	Mon Oct 09 04:35:09 2017
++++ b/folly/Malloc.h	Fri Oct 13 15:31:36 2017
+@@ -189,9 +189,6 @@
+       return false;
+     }
+ 
+-    /* Avoid optimizing away the malloc.  */
+-    asm volatile("" ::"m"(ptr) : "memory");
+-
+     return (origAllocated != *counter);
+   }();
+ 

--- a/ports/folly/fix-static-linking.diff
+++ b/ports/folly/fix-static-linking.diff
@@ -1,0 +1,98 @@
+diff --git a/CMake/folly-deps.cmake b/CMake/folly-deps.cmake
+index 211c6fbf..fbcbddf0 100755
+--- a/CMake/folly-deps.cmake
++++ b/CMake/folly-deps.cmake
+@@ -10,9 +10,26 @@ find_package(Boost 1.55.0 MODULE
+     thread
+   REQUIRED
+ )
++
+ find_package(DoubleConversion MODULE REQUIRED)
+-find_package(GFlags MODULE REQUIRED)
+-find_package(GLog MODULE REQUIRED)
+-find_package(LibEvent MODULE REQUIRED)
++
++find_package(gflags CONFIG)
++if(NOT TARGET gflags)
++  message(STATUS "MODULE: GFlags")
++  find_package(GFlags MODULE REQUIRED)
++endif()
++
++find_package(glog CONFIG)
++if(NOT TARGET glog::glog)
++  message(STATUS "MODULE: GLog")
++  find_package(GLog MODULE REQUIRED)
++endif()
++
++find_package(Libevent CONFIG)
++if(NOT TARGET event)
++  message(STATUS "MODULE: Libevent")
++  find_package(LibEvent MODULE REQUIRED)
++endif()
++
+ find_package(OpenSSL MODULE REQUIRED)
+ find_package(PThread MODULE)
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index cdc992a8..c0231402 100755
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -141,9 +141,38 @@ set(FOLLY_SHINY_DEPENDENCIES
+ 
+ set(FOLLY_LINK_LIBRARIES
+   ${DOUBLE_CONVERSION_LIBRARY}
+-  ${LIBEVENT_LIB}
+-  ${LIBGFLAGS_LIBRARY}
+-  ${LIBGLOG_LIBRARY}
++)
++
++set(FOLLY_INCLUDE_DIRECTORIES
++  ${DOUBLE_CONVERSION_INCLUDE_DIR}
++)
++
++if(TARGET gflags_static)
++  set(FOLLY_SHINY_DEPENDENCIES ${FOLLY_SHINY_DEPENDENCIES} gflags_static)
++elseif(TARGET gflags)
++  set(FOLLY_SHINY_DEPENDENCIES ${FOLLY_SHINY_DEPENDENCIES} gflags)
++else()
++  set(FOLLY_LINK_LIBRARIES ${FOLLY_LINK_LIBRARIES} ${LIBGFLAGS_LIBRARY})
++  set(FOLLY_INCLUDE_DIRECTORIES ${FOLLY_INCLUDE_DIRECTORIES} ${LIBGFLAGS_INCLUDE_DIR})
++endif()
++
++if(TARGET glog::glog)
++  set(FOLLY_SHINY_DEPENDENCIES ${FOLLY_SHINY_DEPENDENCIES} glog::glog)
++else()
++  set(FOLLY_LINK_LIBRARIES ${FOLLY_LINK_LIBRARIES} ${LIBGLOG_LIBRARY})
++  set(FOLLY_INCLUDE_DIRECTORIES ${FOLLY_INCLUDE_DIRECTORIES} ${LIBGLOG_INCLUDE_DIR})
++endif()
++
++if(TARGET event)
++  set(FOLLY_SHINY_DEPENDENCIES ${FOLLY_SHINY_DEPENDENCIES} event)
++else()
++  set(FOLLY_LINK_LIBRARIES ${FOLLY_LINK_LIBRARIES} ${LIBEVENT_LIB})
++  set(FOLLY_INCLUDE_DIRECTORIES ${FOLLY_INCLUDE_DIRECTORIES} ${LIBEVENT_INCLUDE_DIR})
++endif()
++
++
++set(FOLLY_LINK_LIBRARIES
++  ${FOLLY_LINK_LIBRARIES}
+   Iphlpapi.lib
+   Ws2_32.lib
+ 
+@@ -152,15 +181,12 @@ set(FOLLY_LINK_LIBRARIES
+ 
+ target_include_directories(folly_base
+   PUBLIC
+-    ${DOUBLE_CONVERSION_INCLUDE_DIR}
+-    ${LIBGFLAGS_INCLUDE_DIR}
+-    ${LIBGLOG_INCLUDE_DIR}
+-    ${LIBEVENT_INCLUDE_DIR}
+-    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
++    ${FOLLY_INCLUDE_DIRECTORIES}
+ )
+ 
+ foreach (LIB ${FOLLY_SHINY_DEPENDENCIES})
+   target_include_directories(folly_base PUBLIC $<TARGET_PROPERTY:${LIB},INCLUDE_DIRECTORIES>)
++  target_compile_definitions(folly_base PUBLIC $<TARGET_PROPERTY:${LIB},INTERFACE_COMPILE_DEFINITIONS>)
+ endforeach()
+ 
+ if (FOLLY_HAVE_PTHREAD)

--- a/ports/folly/portfile.cmake
+++ b/ports/folly/portfile.cmake
@@ -2,6 +2,10 @@ if(NOT VCPKG_TARGET_ARCHITECTURE STREQUAL x64)
   message(FATAL_ERROR "Folly only supports the x64 architecture.")
 endif()
 
+if(VCPKG_LIBRARY_LINKAGE STREQUAL static)
+  message(FATAL_ERROR "Folly can't link with static libraries.")
+endif()
+
 if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
     message(STATUS "Warning: Dynamic building not supported yet. Building static.")
     set(VCPKG_LIBRARY_LINKAGE static)
@@ -17,9 +21,18 @@ set(ENV{PATH} "$ENV{PATH};${PYTHON3_DIR}")
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO facebook/folly
-    REF v2017.07.17.01
-    SHA512 1529941ebcc9ee3502e4ab07463bb5bc6a397259cbbf1817ac64fd186fd4cd964ffe8b473fe6a432e80632ffa553106a0d3fe2962e860eb409e3391029ed5584
+    REF v2017.10.02.00
+    SHA512 4fc7840c7a6c528c8ba6a21817bc75f15f5cd5f781d104a1f0622fe1085a6cb26ff9749616b164afff0ea46be6d16877457a98f417e6dbe1044db7605650a6d3
     HEAD_REF master
+)
+
+vcpkg_apply_patches(
+    SOURCE_PATH
+        ${SOURCE_PATH}
+    PATCHES
+        ${CMAKE_CURRENT_LIST_DIR}/fix-malloc.diff
+        ${CMAKE_CURRENT_LIST_DIR}/fix-MSG_ERRQUEUE.diff
+        ${CMAKE_CURRENT_LIST_DIR}/fix-histogram.diff
 )
 
 if(VCPKG_CRT_LINKAGE STREQUAL static)

--- a/ports/folly/portfile.cmake
+++ b/ports/folly/portfile.cmake
@@ -2,10 +2,6 @@ if(NOT VCPKG_TARGET_ARCHITECTURE STREQUAL x64)
   message(FATAL_ERROR "Folly only supports the x64 architecture.")
 endif()
 
-if(VCPKG_LIBRARY_LINKAGE STREQUAL static)
-  message(FATAL_ERROR "Folly can't link with static libraries.")
-endif()
-
 if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
     message(STATUS "Warning: Dynamic building not supported yet. Building static.")
     set(VCPKG_LIBRARY_LINKAGE static)
@@ -30,6 +26,7 @@ vcpkg_apply_patches(
     SOURCE_PATH
         ${SOURCE_PATH}
     PATCHES
+        ${CMAKE_CURRENT_LIST_DIR}/fix-static-linking.diff
         ${CMAKE_CURRENT_LIST_DIR}/fix-malloc.diff
         ${CMAKE_CURRENT_LIST_DIR}/fix-MSG_ERRQUEUE.diff
         ${CMAKE_CURRENT_LIST_DIR}/fix-histogram.diff

